### PR TITLE
KAFKA-5036: hold onto the leader lock in Partition while serving an O…

### PR DESCRIPTION
**Pull request by junrao**
…ffsetForLeaderEpoch request

### CODE REVIEW

**Comment by junrao** [comment]
@benstopford @ijuma : Could you review this? Thanks.

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 1]
```diff
@@ -510,6 +512,21 @@ class Partition(val topic: String,
...
+    * @param leaderEpoch Requested leader epoch
+    * @return The last offset of messages published under this leader epoch.
+    */
+  def lastOffsetForLeaderEpoch(leaderEpoch: Int): EpochEndOffset = {
+    inReadLock(leaderIsrUpdateLock) {
+      leaderReplicaIfLocal match {
+        case Some(leaderReplica) =>
+          new EpochEndOffset(NONE, leaderReplica.epochs.get.endOffsetFor(leaderEpoch))
+        case None =>
+          new EpochEndOffset(NOT_LEADER_FOR_PARTITION, UNDEFINED_EPOCH_OFFSET)
```
Comment (ijuma) [inline core/src/main/scala/kafka/cluster/Partition.scala:525]: Nit: `map` followed by `getOrElse` is a little more concise, but OK like this if you prefer. Same applies for `lastOffsetForLeaderEpoch`.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 2]
```diff
@@ -20,36 +20,32 @@ import java.io.{File, IOException}
...
 import kafka.server.QuotaFactory.UnboundedQuota
 import kafka.server.checkpoints.OffsetCheckpointFile
 import kafka.utils._
-import org.apache.kafka.common.errors.{ControllerMovedException, CorruptRecordException, InvalidTimestampException, InvalidTopicException, NotEnoughReplicasException, NotLeaderForPartitionException, OffsetOutOfRangeException, PolicyViolationException}
 import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.errors.{ControllerMovedException, CorruptRecordException, InvalidTimestampException, InvalidTopicException, NotEnoughReplicasException, NotLeaderForPartitionException, OffsetOutOfRangeException, PolicyViolationException, _}
+import org.apache.kafka.common.internals.Topic
 import org.apache.kafka.common.metrics.Metrics
 import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.protocol.Errors._
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: Nit: is it really worth importing this with a wildcard? It seems like it's clearer if we just use the `Errors` prefix. There are only 3 instances where we rely on this in the file.

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 3]
```diff
@@ -16,28 +16,30 @@
...
 import java.util.concurrent.locks.ReentrantReadWriteLock
 
-import org.apache.kafka.common.errors.{PolicyViolationException, NotEnoughReplicasException, NotLeaderForPartitionException}
-import org.apache.kafka.common.protocol.Errors
-
-import scala.collection.JavaConverters._
 import com.yammer.metrics.core.Gauge
+import kafka.admin.AdminUtils
+import kafka.api.LeaderAndIsr
+import kafka.common._
```
Comment (ijuma) [inline core/src/main/scala/kafka/cluster/Partition.scala]: It seems like we can replace this wildcard by simply:

`import kafka.common.NotAssignedReplicaException`

File: core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala [hunk 4]
```diff
@@ -180,7 +180,7 @@ class LeaderEpochFileCache(topicPartition: TopicPartition, leo: () => LogOffsetM
     checkpoint.write(epochs)
   }
 
-  def epochChangeMsg(epoch: Int, offset: Long) = s"New: {epoch:$epoch, offset:$offset}, Latest: {epoch:$latestEpoch, offset$latestOffset} for Partition: $topicPartition"
+  def epochChangeMsg(epoch: Int, offset: Long) = s"New: {epoch:$epoch, offset:$offset}, Old: {epoch:$latestEpoch, offset$latestOffset} for Partition: $topicPartition"
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala]: Nit: it's a little confusing that we log `Old`, but the code is printing the output of `latestEpoch`. Would `Current` perhaps be clearer? I don't feel strongly, so OK to leave it as in the PR if you prefer.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, looks good overall, left a few minor suggestions.

File: core/src/test/scala/unit/kafka/server/epoch/OffsetsForLeaderEpochTest.scala [hunk 5]
```diff
@@ -16,83 +16,90 @@
...
+    val response = replicaManager.lastOffsetForLeaderEpoch(request)
 
     //Then
-    assertEquals(new EpochEndOffset(Errors.NONE, offset), response.get(tp))
+    assertEquals(new EpochEndOffset(Errors.NONE, offset), response(tp))
   }
 
   @Test
   def shonuldReturnNoLeaderForPartitionIfThrown(): Unit = {
-    val replicaManager = createNiceMock(classOf[kafka.server.ReplicaManager])
```
Comment (benstopford) [inline core/src/test/scala/unit/kafka/server/epoch/OffsetsForLeaderEpochTest.scala]: typo (not you but if you're there) 
shonuldReturnNoLeaderForPartitionIfThrown => shouldReturnNoLeaderForPartition

**Review by benstopford — COMMENTED** [review]
LGTM @junrao noticed a typo in the original code, but not a big deal.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 1 above)
Comment (junrao) [reply core/src/main/scala/kafka/cluster/Partition.scala:525]: I left it as it is since the more verbose code seems a bit clearer.

**Review by ijuma — APPROVED** [review]
LGTM

**Pull request closed without merging**